### DISCUSSION TITLE
Named service constructor injection

### DISCUSSION THREAD

**Discussion by lucacivale** [dadhi/DryIoc#665] [Q&A] [answered]
Let's assume I have 
```
interface IService
{
}

class ServiceA : IService
{
}

internal class ServiceB : IService
{
}

interface IOtherService
{
}

class OtherServiceA : IOtherService
{
    public OtherServiceA(IService serviceA)
    {
    }
}

class OtherServiceB : IOtherService
{
    public OtherServiceB(IService serviceB)
    {
    }
}
```
Afaik i have to register the services like this to enable named service constructor injection
```
container.Register<IService, ServiceA>("ServiceA");
container.Register<IService, ServiceB>("ServiceB");

container.Register<IOtherService, OtherServiceA>(
    serviceKey: "OtherServiceA",
    made: Parameters.Of
        .Name("serviceA", serviceKey: "ServiceA"));

container.Register<IOtherService, OtherServiceB>(
    serviceKey: "OtherServiceB",
    made: Parameters.Of
        .Name("serviceB", serviceKey: "ServiceB"));
```

Is there any other way to achieve this? I would like to avoid hardcoded parameter names.
ChatGPT suggested something like this but which looks pretty nice but is kinda messed up

1) Create a helper attribute
```
[AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false)]
public class ServiceKeyAttribute : Attribute
{
    public object Key { get; }

    public ServiceKeyAttribute(object key)
    {
        Key = key;
    }
}
```

2) Use it in your service
```
public class MyService
{
    private readonly IDependency _dependency;

    public MyService([ServiceKey("MyNamedService")] IDependency dependency)
    {
        _dependency = dependency;
    }
}
```

3) Configure the DryIoc Container
```
var rules = DryIocContainerExtension.GetDefaultRules()
    .WithFactoryMethodInjection()
    .With(FactoryMethod.InjectAttribute<ServiceKeyAttribute>(attr => attr.Key)));
```

**Thank you in advance for your help!**

**Comment by dadhi** [comment]
@lucacivale If you are using MS.DI and DryIoc.Microsoft.DependencyInjection, then you may utilize `[FromKeyedServices]` attribute:
```cs
        internal class OtherService
        {
            public OtherService(
                [FromKeyedServices("service1")] IService service1,
                [FromKeyedServices("service2")] IService service2)
            {
                Service1 = service1;
                Service2 = service2;
            }

            public IService Service1 { get; }
            public IService Service2 { get; }
        }
   ```
   
For this, you need https://www.nuget.org/packages/DryIoc.Microsoft.DependencyInjection/8.0.0-preview-03 and DryIoc.dll v6.0.0-preview-08

Or you may use DryIoc.MefAttributedModel.dll and use a usual MEF Export, Import attributes.

**Comment by lucacivale** [comment]
@dadhi thanks for the quick response. I'm using the Prism library including the Prism.Container.DryIoc package wich references DryIoc.dll. Do you know if this should also work or do I have to work with Microsoft.DependencyInjection

**Reply by dadhi** [reply, marked as answer]
@lucacivale I am not sure about Prism. I would've asked the maintainer if and where I can add `.WithMefAttributedModel()` to the container configuration. 
I think adding package dependency of DryIoc.MefAttributedModel.dll should not be an issue.

**Comment by lucacivale** [comment]
@dadhi thank you, works like a charm
